machine/parser.go

```diff
@@ -27,67 +27,35 @@ func FromJSON(raw []byte) (*StateMachine, error) {
 	return &sm, err
 }
 
-type basicMachine struct {
-	Comment *string
-	StartAt *string
-}
-
-// from http://gregtrowbridge.com/golang-json-serialization-with-interfaces/
-func (sm *StateMachine) UnmarshalJSON(b []byte) error {
-	// First, deserialize StateMachine fields
-	var bsm basicMachine
-	err := json.Unmarshal(b, &bsm)
-	if err != nil {
-		return err
-	}
-	sm.Comment = bsm.Comment
-	sm.StartAt = bsm.StartAt
-
+func (sm *States) UnmarshalJSON(b []byte) error {
 	// States
-	var objMap map[string]*json.RawMessage
-	err = json.Unmarshal(b, &objMap)
+	var rawStates map[string]*json.RawMessage
+	err := json.Unmarshal(b, &rawStates)
```

Comment:
Unless I am misreading this, this seems backwards incompatible. The previous version looks like it was parsing a JSON with a States key, such as (from parser_test.go), and reaches into the object under the States key, and unmarshalls each state in a for loop.
```
{
      "States": {
        "Coords": {
          "Type": "Pass",
          "Result": {
            "x": 3.14,
            "y": 103.14159
          },
          "ResultPath": "$.coords",
          "End": true
        }
    }
  }
```

However, the new version, doesnt even reach into the "States" key, it (from my understanding) tries to do a for loop on the base level JSON object instead, so I don't think this works.

But CircleCI seems to pass 🤷‍♂️

Reply:
I am parsing the States collection now. This is a trick I found that if you want to parse a list of interfaces you can declare a type, e.g. `type States map[string]interface{}` and then write a parser directly for the collection, so you don't have to parse the outer struct.

Reply:
Gotcha - just realized there is no test for this function, can you write one so we can be sure it works? I don't see one, would be helpful

Reply:
This function is called by `json.Unmarshal()` which is use in most tests.

Reply:
Oh! I didn't know that, this makes more sense now, thanks for the clarification